File: 49-group-anagrams/49-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        
        for (String word : strs) {
            //string is immutable, to sort word, it has to be converted to charArray then convert back
            char[] charArray = word.toCharArray();
            Arrays.sort(charArray);
            String sortedWord = new String(charArray);
            
            if (map.containsKey(sortedWord)) {
                map.get(sortedWord).add(word);
            }
            else {
                map.put(sortedWord, new ArrayList<String>                                                               (Arrays.asList(word)));
            }
        }
        return new ArrayList<List<String>>(map.values());
    }
}

File: 49-group-anagrams/49-group-anagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        hashTable = {}
        for word in strs:
            sortedWord = "".join(sorted(word))
            if sortedWord in hashTable:
                hashTable[sortedWord].append(word)
            else:
                hashTable[sortedWord] = [word]
        return list(hashTable.values())